### PR TITLE
Added ECDSA; Added RSA tests; Fixed linting errors; Handling all un-handled errors

### PR DESCRIPTION
**Pull request by adam-hanna**
1. [Added ECDSA](https://github.com/c3systems/go-libp2p-crypto/blob/master/ecdsa.go)
2. [Added RSA tests](https://github.com/c3systems/go-libp2p-crypto/blob/master/rsa_test.go)
3. Fixed linting errors. Mostly just meant adding comments to exported types and functions
4. Handled all un-handled errors. Please note that this is not backwards compatible as I had to add an error in the return of the `KeyStretcher` function. Every time `m.Write(..)` gets called (for example, [here](https://github.com/libp2p/go-libp2p-crypto/blob/master/key.go#L200)), the number of bytes written and an error gets returned. These returned values were just being ignored.

### CODE REVIEW

**Comment by adam-hanna** [comment]
I think the build is failing because it's pulling `github.com/libp2p/go-libp2p-crypto/pb` which doesn't have the required types, rather than using the types in the included subdir.

**Comment by adam-hanna** [comment]
Merged [#34](https://github.com/libp2p/go-libp2p-crypto/pull/34) and [#37 ](https://github.com/libp2p/go-libp2p-crypto/pull/37)

File: key.go [hunk 1]
```diff
@@ -207,15 +219,21 @@ func KeyStretcher(cipherType string, hashType string, secret []byte) (StretchedK
 	}
 
 	m := hmac.New(h, secret)
-	m.Write(seed)
+	if _, err := m.Write(seed); err != nil {
```
Comment (Stebalien) [inline key.go]: This is guaranteed to never return an error.

File: key.go [hunk 2]
```diff
@@ -163,15 +174,16 @@ func GenerateEKeyPair(curveName string) ([]byte, GenSharedKey, error) {
...
 	IV        []byte
 	MacKey    []byte
 	CipherKey []byte
 }
 
-// Generates a set of keys for each party by stretching the shared key.
+// KeyStretcher returns a set of keys for each party by stretching the shared key.
 // (myIV, theirIV, myCipherKey, theirCipherKey, myMACKey, theirMACKey)
-func KeyStretcher(cipherType string, hashType string, secret []byte) (StretchedKeys, StretchedKeys) {
+func KeyStretcher(cipherType string, hashType string, secret []byte) (*StretchedKeys, *StretchedKeys, error) {
```
Comment (Stebalien) [inline key.go]: This is a public interface so we probably shouldn't change it (unless there's a very good reason).

**Comment by Stebalien** [comment]
While you're at it, could you rebase on master and cleanup the git history a bit? Reverse merges get really messy and it would be nice to have all of the documentation/cleanup (thanks, by the way) changes in separate commits.

File: key.go (hunk 2 above)
Comment (Stebalien) [reply key.go]: See my comment below (above?). Writes to hmac's can't fail (according to the documentation). We should probably have a comment stating this...

Comment (adam-hanna) [reply key.go]: My apologies! Didn't see your other comment and didn't know that about HMAC writes! TY!

I'll leave this function definition unchanged, then.

Comment (Stebalien) [reply key.go]: Np, no need to apologize. This is definitely non-obvious (I had to go check the docs thinking "we can't have been *that* lazy...").

File: secp256k1.go [hunk 3]
```diff
@@ -40,27 +46,32 @@ func UnmarshalSecp256k1PublicKey(data []byte) (PubKey, error) {
...
 
+// Raw returns the bytes of the key
 func (k *Secp256k1PrivateKey) Raw() ([]byte, error) {
 	return (*btcec.PrivateKey)(k).Serialize(), nil
 }
 
+// Equals compares two private keys
 func (k *Secp256k1PrivateKey) Equals(o Key) bool {
-	sk, ok := o.(*Secp256k1PrivateKey)
+	k, ok := o.(*Secp256k1PrivateKey)
```
Comment (Stebalien) [inline secp256k1.go]: Bug.

File: secp256k1.go [hunk 4]
```diff
@@ -71,31 +82,37 @@ func (k *Secp256k1PrivateKey) Sign(data []byte) ([]byte, error) {
...
 
+// Raw returns the bytes of the key
 func (k *Secp256k1PublicKey) Raw() ([]byte, error) {
 	return (*btcec.PublicKey)(k).SerializeCompressed(), nil
 }
 
+// Equals compares two public keys
 func (k *Secp256k1PublicKey) Equals(o Key) bool {
-	sk, ok := o.(*Secp256k1PublicKey)
+	k, ok := o.(*Secp256k1PublicKey)
```
Comment (Stebalien) [inline secp256k1.go]: Bug!

File: key.go [hunk 5]
```diff
@@ -5,55 +5,65 @@ package crypto
...
+	"encoding/base64"
+	"errors"
+	"fmt"
 	"hash"
-
-	pb "github.com/libp2p/go-libp2p-crypto/pb"
+	"io"
 
 	"github.com/gogo/protobuf/proto"
+	pb "github.com/libp2p/go-libp2p-crypto/pb"
```
Comment (Stebalien) [inline key.go]: FYI, we usually order our imports:

1. stdlib
2. local repo
3. external repos

(with newlines in-between)

However, the separation between the crypto and non-crypto imports was *non*-idiomatic.

File: ecdsa.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	priv, err := ecdsa.GenerateKey(curve, src)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &ECDSAPrivateKey{priv}, &ECDSAPublicKey{&priv.PublicKey}, nil
+}
+
+// GenerateECDSAKeyPairFromKey generates a new ecdsa private and public key from an input private key
+func GenerateECDSAKeyPairFromKey(priv *ecdsa.PrivateKey) (PrivKey, PubKey, error) {
```
Comment (Stebalien) [inline ecdsa.go]: Maybe drop the `Generate` (we're not really generating a new key)?

File: ecdsa.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	ErrNilSig = errors.New("sig is nil")
+	// ErrNilPrivateKey is returned when a nil private key is provided
+	ErrNilPrivateKey = errors.New("private key is nil")
+	// ECDSACurve is the default ecdsa curve used
+	ECDSACurve = elliptic.P256()
+)
+
+// GenerateECDSAKeyPair generates a new ecdsa private and public key
+func GenerateECDSAKeyPair(src io.Reader) (PrivKey, PubKey, error) {
+	priv, err := ecdsa.GenerateKey(ECDSACurve, src)
```
Comment (Stebalien) [inline ecdsa.go]: Can replace with `return GenerateECDSAKeyPairWithCurve(ECDSACurve, src)`.

**Comment by adam-hanna** [comment]
@Stebalien thanks, I'm on it!

**Comment by adam-hanna** [comment]
@Stebalien, thanks again for all of your help, and sorry for introducing that bug. Great catch!

I just incorporated your latest round of feedback...

**Comment by Stebalien** [comment]
I'd like to get @dignifiedquire to take a look at this before merging as he maintains the JS version. I also noticed that "secp256k1" is *also* ECDSA (with a specific curve, maybe some bitcoin-specific key formats?). Hopefully @dignifiedquire will know more.

**Comment by Stebalien** [comment]
I've discussed this and, well, we *do* support a variant of ECDSA (Secp256k1) but we use a hard-coded curve and a minimal key format. The upside is that keys in *that* format are small. The downside, is that it's a bit less flexible. The downside of *this* format is that keys are large. However, it supports all "standard" keys so it should help with interoperability with existing systems.

Given that, we've (@dignifiedquire and I) have decided to merge this.

**Comment by Stebalien** [comment]
Thanks @adam-hanna for bearing with this!

**Comment by adam-hanna** [comment]
Thanks for everything!

File: package.json [hunk 8]
```diff
@@ -7,12 +7,6 @@
     "dvcsimport": "github.com/libp2p/go-libp2p-crypto"
   },
   "gxDependencies": [
-    {
-      "author": "whyrusleeping",
-      "hash": "QmdxUuburamoF6zF9qjeQC4WYcWGbWuRmdLacMEsW8ioD8",
-      "name": "gogo-protobuf",
-      "version": "0.0.0"
-    },
```
Comment (Stebalien) [inline package.json]: Why?

**Pull request merged**